File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
       unordered_map<int,int> mp;

       for(int i = 0;i<inorder.size();i++) {
        mp[inorder[i]] = i;
       }

       int ps = 0;
       return dfs(preorder, inorder, mp, 0, inorder.size()-1, ps);
    }

    TreeNode* dfs(vector<int>& pre, vector<int>& ino, unordered_map<int,int>& mp, int l, int r, int &p) {
        if(l>r) return nullptr;

        TreeNode* root = new TreeNode(pre[p]);
        int inPo = mp[pre[p++]];

        root->left = dfs(pre, ino, mp, l, inPo-1, p);
        root->right = dfs(pre, ino, mp, inPo+1, r, p);

        return root;
    }

};

File: Data Structures & Algorithms/count-subsequences/submission-2.java
class Solution {
    public int numDistinct(String s, String t) {
        int n = s.length(), m = t.length();
        int[][] dp = new  int[n+1][m+1];
        for(int i = 0;i<=n;i++) {
            dp[i][0] = 1;
        }

        for(int i = 1;i<=n;i++) {
            for(int j = 1;j<=m;j++) {
                if(s.charAt(i-1) == t.charAt(j-1)) {
                    dp[i][j] += dp[i-1][j-1];
                }
                dp[i][j] += dp[i-1][j];
            }
        }
        return dp[n][m];
        // return dfs(0,0,s,t, dp);
    }

    public int dfs(int i, int j, String s, String t, int[][] dp) {
        int n = s.length(), m = t.length();
        if(j>=m) {
            return 1;
        }
        if(i>=n){
            return 0;
        }

        if(dp[i][j] != -1) return dp[i][j];

        int ans = 0;

        if(s.charAt(i) == t.charAt(j)){
            ans+=dfs(i+1, j+1,s,t, dp);
        }
        ans += dfs(i+1, j, s, t, dp);

        dp[i][j] = ans;
        return ans;
    }
}
